### PR TITLE
Hostremaster

### PR DESCRIPTION
**Pull request by mhasself**
HostMaster revamp

(will replace this when ready for merge)

Goals are:
- Facilitate control docker containers through HostMaster running on the host system
- Make web widget better
- Bring up to date with modern OCS stuff

### CODE REVIEW

**Comment by mhasself** [comment]
Ready for review.  Ok, so there aren't enough tests.  But ready other than that.

File: agents/host_manager/host_manager.py
Comment (BrianJKoopman) [inline agents/host_manager/host_manager.py:209]: Do these get left up? It appears if I startup a bunch of containers with an HM, kill the HM, then startup the HM again it detects these active containers, then requests launches for them and states they are launched.

```
2022-01-27T19:42:48+0000 startup-op: launching manager
2022-01-27T19:42:48+0000 start called for manager
2022-01-27T19:42:48+0000 manager:0 Status is now "starting".
2022-01-27T19:42:48+0000 manager:0 Status is now "running".
2022-01-27T19:42:49+0000 manager:0 Loaded 3 agent instance configs.
2022-01-27T19:42:49+0000 manager:0 On startup, detected active container for ocs-client
2022-01-27T19:42:49+0000 manager:0 On startup, detected active container for fake-data1
2022-01-27T19:42:49+0000 manager:0 On startup, detected active container for ocs-registry
2022-01-27T19:42:50+0000 manager:0 Requested launch for docker:ocs-client
2022-01-27T19:42:50+0000 manager:0 Requested launch for docker:fake-data1
2022-01-27T19:42:50+0000 manager:0 Requested launch for docker:ocs-registry
2022-01-27T19:42:52+0000 manager:0 Launched docker:ocs-client
2022-01-27T19:42:52+0000 manager:0 Launched docker:fake-data1
2022-01-27T19:42:52+0000 manager:0 Launched docker:ocs-registry
```

**Review by BrianJKoopman — COMMENTED** [review]
This is awesome! I have some thoughts on the current configuration recommendations.

- When I read through the "Centralized Management" page I got the impression that if I need to control both Agents running on the native host and Agents running in Docker containers I needed two Host Managers to do so. As we talked about on Slack, this doesn't have to be the case, but I think we need to recommend a singular HM on the native host, and not recommend putting an HM in the `<host>-docker` block of a SCF for a few reasons:
  - For the HM Agent to run properly it requires the 'agent-paths' configuration block within the same host config. In the case of being in a Docker config block this would mean pointing to the native host agent-paths, which is potentially confusing to users. For Docker containers we typically omit this block because the Agents being run are in the working directory of the container, so they don't really need it.
  - The `wamp_server` and `wamp_http` URLs are typically different for Agents in the containers and those on the host. And while we can configure these addresses per host, having an HM in the docker host block mixes these types.
  - Managing startup of up this 2nd HM would require a secondary ocs-hostmanager.service file.
- The user that runs the HM Agent needs to be in the 'docker' group. The ocs documentation currently recommends not putting your user in the 'docker' group, but instead running with sudo. I think we need some text on this. The recommendation should be to follow the existing docs on [setting up the ocs user](https://ocs.readthedocs.io/en/develop/user/installation.html#create-ocs-user), and to put that user in the 'docker' group. Possibly we should lock down that account a bit as well. I like this separation, as it means the individual users running OCS don't need 'docker' or 'sudo' permissions. However, this setup does introduce some potential difficulties.
  - If the HM goes down in such a way that the containers remain online, or in a strange state, then it could be difficult for users to resolve the situation by bringing down the containers before restarting the HM Agent. Is it alright if they don't do this? (I thought at one point I had to to get HM to startup successfully.)

Some errors I occasionally ran into:
- When running "down" for a single Agent in a container via the web interface (couldn't reproduce this instance) and when starting up containers by hand once when HM wasn't starting them itself (which I know is not intended behavior, but it produced a similar error):
```
2022-01-27T16:52:04+0000 session joined:
SessionDetails(realm="test_realm",
               session=7889680223690494,
               authid="V7JY-Q3AN-PC74-X33T-EVA5-69NP",
               authrole="iocs_agent",
               authmethod="anonymous",
               authprovider="static",
               authextra={'x_cb_node': '8d9995660cdb-1', 'x_cb_worker': 'worker001', 'x_cb_peer': 'tcp4:172.20.0.1:35494', 'x_cb_pid': 11},
               serializer="msgpack.batched",
               transport="websocket",
               resumed=None,
               resumable=None,
               resume_token=None)
2022-01-27T16:52:04+0000 startup-op: launching manager
2022-01-27T16:52:04+0000 start called for manager
2022-01-27T16:52:04+0000 manager:0 Status is now "starting".
2022-01-27T16:52:04+0000 manager:0 Status is now "running".
2022-01-27T16:52:05+0000 manager:0 Loaded 3 agent instance configs.
2022-01-27T16:52:06+0000 manager:0 Requested launch for docker:ocs-client
2022-01-27T16:52:06+0000 manager:0 Requested launch for docker:fake-data1
2022-01-27T16:52:06+0000 manager:0 Requested launch for docker:ocs-registry
2022-01-27T16:52:09+0000 manager:0 Launched docker:ocs-client
2022-01-27T16:52:09+0000 manager:0 Launched docker:fake-data1
2022-01-27T16:52:09+0000 manager:0 Launched docker:ocs-registry
2022-01-27T16:57:25+0000 start called for update
2022-01-27T16:57:25+0000 update:1 Status is now "starting".
2022-01-27T16:57:26+0000 update:1 Update requested.
2022-01-27T16:57:26+0000 update:1 Status is now "done".
2022-01-27T16:57:26+0000 update:1 Loaded 3 agent instance configs.
2022-01-27T16:57:26+0000 manager:0 Requesting termination of docker:fake-data1
2022-01-27T16:57:28+0000 manager:0 CRASH: [Failure instance: Traceback: <class 'RuntimeError'>: Trouble running "docker inspect df77f43b4866491f93d9e5567750c0da6471af3de7024110531937a5e025c011".
/usr/lib/python3/dist-packages/twisted/internet/defer.py:501:errback
/usr/lib/python3/dist-packages/twisted/internet/defer.py:568:_startRunCallbacks
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1475:gotResult
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1416:_inlineCallbacks
/usr/lib/python3/dist-packages/twisted/python/failure.py:491:throwExceptionIntoGenerator
/home/vagrant/git/ocs/agents/host_manager/host_manager.py:311:manager
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1416:_inlineCallbacks
/usr/lib/python3/dist-packages/twisted/python/failure.py:491:throwExceptionIntoGenerator
/home/vagrant/git/ocs/agents/host_manager/host_manager.py:81:_update_docker_states
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1418:_inlineCallbacks
/home/vagrant/git/ocs/ocs/agent/host_manager.py:309:parse_docker_state
]
2022-01-27T16:57:28+0000 manager:0 Status is now "done".
```
- I see this one fairly frequently, which looks like it matches what happens if "manage:" is typo'd, though I usually can't reproduce it other times I see it:
```
2022-01-27T19:50:00+0000 startup-op: launching manager
2022-01-27T19:50:00+0000 start called for manager
2022-01-27T19:50:00+0000 manager:0 Status is now "starting".
2022-01-27T19:50:00+0000 manager:0 Status is now "running".
2022-01-27T19:50:01+0000 manager:0 CRASH: [Failure instance: Traceback: <class 'KeyError'>: ('docker', 'ocs-client')
/usr/lib/python3/dist-packages/twisted/internet/defer.py:501:errback
/usr/lib/python3/dist-packages/twisted/internet/defer.py:568:_startRunCallbacks
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1475:gotResult
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1416:_inlineCallbacks
/usr/lib/python3/dist-packages/twisted/python/failure.py:491:throwExceptionIntoGenerator
/home/vagrant/git/ocs/agents/host_manager/host_manager.py:318:manager
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1418:_inlineCallbacks
/home/vagrant/git/ocs/agents/host_manager/host_manager.py:86:_update_docker_states
]
2022-01-27T19:50:01+0000 manager:0 Status is now "done".
2022-01-27T19:50:01+0000 manager:0 Loaded 3 agent instance configs.
2022-01-27T19:50:01+0000 manager:0 On startup, detected active container for ocs-client
2022-01-27T19:50:01+0000 manager:0 On startup, detected active container for fake-data1
2022-01-27T19:50:01+0000 manager:0 On startup, detected active container for ocs-registry
```

I didn't look at ocsbow or the HostManager.js very closely, but those both worked for me during testing.

Also, thank you for adding some tests!

File: docs/user/centralized_management.rst
Comment (mhasself) [reply docs/user/centralized_management.rst]: Changed bracketed to: "(Note that in the case of Agents in docker containers, the HostManager will use docker and docker-compose to monitor the state of containers, and request start or stop in order to match the target state.)"

File: tests/integration/test_host_manager_agent_integration.py
Comment (mhasself) [reply tests/integration/test_host_manager_agent_integration.py:57]: Oh, weird ... removed 50-53 in favor of this polling loop.  I think it did happen at least once that it took more than 1s to come up.

File: tests/integration/test_host_manager_agent_integration.py
Comment (mhasself) [reply tests/integration/test_host_manager_agent_integration.py]: Fixing with a merge from develop.

File: docs/user/centralized_management.rst
Comment (mhasself) [reply docs/user/centralized_management.rst:144]: I am not sure why it would require agent-paths.  It doesn't, in my test case.  Agent-paths shouldn't matter unless the HostManager needs to launch an Agent instance (rather than docker-compose something), which it should not do if manage:docker is in all the agent configs.

But ... agreed that this recommendation should be changed so the HM is not defined in the docker host config block.  Then manage:docker won't matter.

File: docs/user/centralized_management.rst
Comment (mhasself) [reply docs/user/centralized_management.rst:300]: This looks as expected; you have a service for ocs-client but no entry in site-config.  The ? ? entry comes from having an entry in the site config that is marked as manage: docker, because those instances can't be _directly_ monitored by HostManager (rather, it can only monitor the docker containers that probably are hosting it).

File: agents/host_manager/host_manager.py
Comment (mhasself) [reply agents/host_manager/host_manager.py:209]: Logging was misleading ... fixed in 0e7ac6.  Won't show the Requested and Launched lines.

This must have been with --initial-state=up?  Without that switch, HM would stop those containers.

**Comment by mhasself** [comment]
> When I read through the "Centralized Management" page I got the impression that if I need to control both Agents running on the native host and Agents running in Docker containers I needed two Host Managers to do so. As we talked about on Slack, this doesn't have to be the case, but I think we need to recommend a singular HM on the native host, and not recommend putting an HM in the `<host>-docker` block of a SCF for a few reasons:
> 
>     * For the HM Agent to run properly it requires the 'agent-paths' configuration block within the same host config. In the case of being in a Docker config block this would mean pointing to the native host agent-paths, which is potentially confusing to users. For Docker containers we typically omit this block because the Agents being run are in the working directory of the container, so they don't really need it.
> 
>     * The `wamp_server` and `wamp_http` URLs are typically different for Agents in the containers and those on the host. And while we can configure these addresses per host, having an HM in the docker host block mixes these types.
> 
>     * Managing startup of up this 2nd HM would require a secondary ocs-hostmanager.service file.

Totally... I've rewritten the instructions based on this.  (fd0dd2)

**Comment by mhasself** [comment]
> * The user that runs the HM Agent needs to be in the 'docker' group. The ocs documentation currently recommends not putting your user in the 'docker' group, but instead running with sudo. I think we need some text on this. The recommendation should be to follow the existing docs on [setting up the ocs user](https://ocs.readthedocs.io/en/develop/user/installation.html#create-ocs-user), and to put that user in the 'docker' group. Possibly we should lock down that account a bit as well. I like this separation, as it means the individual users running OCS don't need 'docker' or 'sudo' permissions. However, this setup does introduce some potential difficulties.

Ok ... see note added in c1188e and dd2b92.

      
>       * If the HM goes down in such a way that the containers remain online, or in a strange state, then it could be difficult for users to resolve the situation by bringing down the containers before restarting the HM Agent. Is it alright if they don't do this? (I thought at one point I had to to get HM to startup successfully.)

The HM can handle that... it will leave them up (if initial state=up) or take them down (otherwise).

> 
> Some errors I occasionally ran into:
> 
>     * When running "down" for a single Agent in a container via the web interface (couldn't reproduce this instance) and when starting up containers by hand once when HM wasn't starting them itself (which I know is not intended behavior, but it produced a similar error):
> 

This must be a race condition, where a container disappeared.  I added some debugging text to the error so we can confirm this is usually the cause.  (And if it is there is probably a good work-around.)

> 
>     * I see this one fairly frequently, which looks like it matches what happens if "manage:" is typo'd, though I usually can't reproduce it other times I see it:

I added a guard around this KeyError ... it shouldn't happen unless docker-compose.yaml is modified while agent is running; now the log will tell you to restart the Agent.

You don't actually have to restart the agent ... there are other things (update task or restarting manage process) that will trigger a rescan that is enough.  We could deal with changes to docker-compose.yaml in a fairly graceful way with a bit of effort.  We should decide what we want to happen in this case.  For now let's say that modifying docker-compose.yaml while the Agent is running leads to undefined behavior ...

**Comment by mhasself** [comment]
@BrianJKoopman I think I've dealt with everything... your move...

File: agents/host_manager/host_manager.py
Comment (BrianJKoopman) [reply agents/host_manager/host_manager.py]: It was with `--initial-state=up`, yes.

File: docs/user/centralized_management.rst
Comment (BrianJKoopman) [reply docs/user/centralized_management.rst]: Right, I wasn't seeing a ? ? entry for the fake-data1 Agent, which has a `manage: docker` line. `ocs-client` isn't an Agent, just another container that I can run ocs clients from. Maybe some indexing was getting messed up by the ocs-client config existing? And the line should have read `fake-data1` for its instance-id instead?

I need to check this again with the latest version.

Comment (mhasself) [reply docs/user/centralized_management.rst]: Ah, sorry!

Looking into this, there is complex and buggy behavior when an instance-id and a service name are the same.  I'll address it.

I think the behavior should be that if you have an instance called X and it's marked with manage: 'docker', _and_ you also have a service named X, then ocs should assume those correspond and merge the configs in a helpful way.

File: agents/host_manager/host_manager.py
Comment (mhasself) [reply agents/host_manager/host_manager.py]: I believe this is resolved, let me know if not.

Comment (BrianJKoopman) [reply agents/host_manager/host_manager.py]: Sorry, ran out of time last night. This looks good to me now.

File: docs/user/centralized_management.rst
Comment (BrianJKoopman) [reply docs/user/centralized_management.rst]: Ah, interesting. This might not be too common in practice, since we usually prefix "ocs-" to the service names. Sounds like a good edge case to handle though.

File: docs/user/centralized_management.rst
Comment (BrianJKoopman) [reply docs/user/centralized_management.rst]: I think it's because otherwise it won't be able to find the host_manager.py file to run. Trying again I get:
```
$ ocs-local-support start agent --foreground --site-host ocs1-docker
Trying to launch hostmanager agent...
Traceback (most recent call last):
  File "/home/vagrant/.local/bin/ocs-local-support", line 11, in <module>
    load_entry_point('ocs', 'console_scripts', 'ocs-local-support')()
  File "/home/vagrant/git/ocs/ocs/ocsbow.py", line 931, in main_local
    raise OcsbowError('Failed to start manager process: %s' % message)
ocs.ocsbow.OcsbowError: Failed to start manager process: Could not find host_manager.py in the agent_paths!
```

Sounds like it won't be a problem now though. The changes to the docs look good.

**Comment by BrianJKoopman** [comment]
> I added a guard around this KeyError ... it shouldn't happen unless docker-compose.yaml is modified while agent is running; now the log will tell you to restart the Agent.

This then runs into:
```
2022-02-09T14-59-13.939246 startup-op: launching manager
2022-02-09T14-59-13.939627 start called for manager
2022-02-09T14-59-13.940118 manager:0 Status is now "starting".
2022-02-09T14-59-13.940694 manager:0 Status is now "running".
2022-02-09T14-59-14.550927 Unexpected service in docker-compose: fake-data1
2022-02-09T14-59-14.551495 Restart the Agent?
2022-02-09T14-59-14.553037 manager:0 CRASH: [Failure instance: Traceback: <class 'UnboundLocalError'>: local variable 'db' referenced before assignment
/usr/lib/python3/dist-packages/twisted/internet/defer.py:501:errback
/usr/lib/python3/dist-packages/twisted/internet/defer.py:568:_startRunCallbacks
/usr/lib/python3/dist-packages/twisted/internet/defer.py:654:_runCallbacks
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1475:gotResult
--- <exception caught here> ---
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1416:_inlineCallbacks
/usr/lib/python3/dist-packages/twisted/python/failure.py:491:throwExceptionIntoGenerator
/home/vagrant/git/ocs/agents/host_manager/host_manager.py:323:manager
/usr/lib/python3/dist-packages/twisted/internet/defer.py:1418:_inlineCallbacks
/home/vagrant/git/ocs/agents/host_manager/host_manager.py:91:_update_docker_states
]
2022-02-09T14-59-14.553954 manager:0 Status is now "done".
2022-02-09T14-59-14.555204 manager:0 Loaded 2 agent instance configs.
```

I'm not sure why it thinks fake-data1 is unexpected though. I think maybe the `hm_utils.parse_docker_state()` call is taking longer sometimes and `services` isn't complete yet?

I can sit there and restart the ocs-hostmanager.service over and over, sometimes I hit it, other times I don't.

My configs at the moment:
```
  ocs1-docker: {

    # List of additional paths to Agent plugin modules.
    #'agent-paths': [
    #  '/home/vagrant/git/ocs/agents/',
    #],
    #'wamp_server': 'ws://localhost:8001/ws',
    #'wamp_http': 'http://localhost:8001/call',


    # Quick start example Agents

    'agent-instances': [
            #{'agent-class': 'HostManager',
            # 'instance-id': 'hm-ocs1-docker',
            # 'arguments': [['--docker-compose', '/home/vagrant/site/docker-compose.yaml']]},
      {'agent-class': 'FakeDataAgent',
       'instance-id': 'fake-data1',
       'manage': 'docker',
       'arguments': [['--mode', 'acq'],
                     ['--num-channels', '16'],
                     ['--sample-rate', '5']]},
      {'agent-class': 'RegistryAgent',
       'instance-id': 'registry',
       'manage': 'docker',
       'arguments': []},
    ]
  }
```

```
services:
        ## Client for fully containerized interactions
        #ocs-client:
        #  #image: socs:latest
        #  image: ocs:latest
        #  stdin_open: true
        #  tty: true
        #  hostname: ocs1-docker
        #  volumes:
        #    - ./:/config:ro
        #    - "./clients:/clients"
        #  environment:
        #    - OCS_CONFIG_DIR=/config
        #  working_dir: /clients

  ocs-fake-data1:
    image: ocs-fake-data-agent:latest
    #image: test-fake-data
    hostname: ocs1-docker
    environment:
      - LOGLEVEL=debug
    volumes:
      - ./:/config:ro
    command:
      - "--instance-id=fake-data1"
      - "--site-hub=ws://crossbar:8001/ws"
      - "--site-http=http://crossbar:8001/call"

  ocs-registry:
      image: ocs-registry-agent:latest
      container_name: ocs-registry
      hostname: ocs1-docker
      user: "9000"
      volumes:
          - ./:/config:ro
      command:
        - "--site-hub=ws://crossbar:8001/ws"
        - "--site-http=http://crossbar:8001/call"
```

**Review by BrianJKoopman — COMMENTED** [review]
> We could deal with changes to docker-compose.yaml in a fairly graceful way with a bit of effort. We should decide what we want to happen in this case. For now let's say that modifying docker-compose.yaml while the Agent is running leads to undefined behavior ...

Yeah, something that occurred to me when I went to update my test system with the latest code on this branch is how we handle updates. Right now we'd change the docker image tags and run `docker-compose up -d`. If the tag changed this'll update that container.

Now I think we'd want to stop the HM Agent, modify the docker-compose file, then start the HM Agent again. Maybe modifying while running then issuing a restart would be alright?

These changes so far look good to me. There's just the conflicting service-name/instance-id thing you were going to address. And now this occasional unexpected docker-service issue described in https://github.com/simonsobs/ocs/pull/234#issuecomment-1033881789.

**Comment by mhasself** [comment]
> Yeah, something that occurred to me when I went to update my test system with the latest code on this branch is how we handle updates. Right now we'd change the docker image tags and run `docker-compose up -d`. If the tag changed this'll update that container.

This all works off the service name, and it just uses docker-compose up -d ... so I think if you use HostManager to stop the agent, then start it again, docker-compose will handle building / getting the update.  Like docker-compose normally would.

> Now I think we'd want to stop the HM Agent, modify the docker-compose file, then start the HM Agent again. Maybe modifying while running then issuing a restart would be alright?

The system is more robust to config changes now.  If you make a syntax error, and save it ... that will cause trouble.  But if you just add or delete a service it's ok.  And if you run the "update" task (with reload_config=True) it should redo everything self-consistently.

> These changes so far look good to me. There's just the conflicting service-name/instance-id thing you were going to address. And now this occasional unexpected docker-service issue described in [#234 (comment)](https://github.com/simonsobs/ocs/pull/234#issuecomment-1033881789).

The latest commits bring a significant overhaul (sorry) to handle same reconciliation of service and instance-id.  So we're back, actually, to recommending that HostManager be listed in the host config where the agents are declared, rather than the native system host config.  This requires some use of --site-host=... argument to agents.  But that's all you need to get around something being in host block that does not match the reported hostname.

File: docs/user/centralized_management.rst
Comment (BrianJKoopman) [inline docs/user/centralized_management.rst:199]: "If HostManager find services..." -> "finds"

File: agents/host_manager/host_manager.py
Comment (BrianJKoopman) [inline agents/host_manager/host_manager.py:134]: These are both assigned, but not used.

**Review by BrianJKoopman — APPROVED** [review]
Sorry for the slow response on this. This looks good to me now. There's one thing I found now that we're back to double host managers (one in the docker block) -- this seems to need two systemd service files now. A few things about this setup:
- I generated a second launcher script with `ocs-install-systemd --service-dir=. --instance-id='hm-ocs1-docker'`. This writes to the same `ocs-hostmanager.service` file. Perhaps this could be unique too?
- A note about needing this second service file in the docs would be good.
- The launcher script generated for this docker launcher had the wrong site-host, using `--site-host=ocs1` when it needed `--site-host=ocs1-docker`.

That said, hand modifying those two things was not a big deal and I got everything working.

Two other super small inline comments. I'd say handle the .service file as you wish, and merge when ready!

**Comment by mhasself** [comment]
> Two other super small inline comments. I'd say handle the .service file as you wish, and merge when ready!

Great -- thanks again!  I added a section to the docs about tweaking ocs-systemd-install output a bit; see final commit.

Will merge if/when test pass.

**Pull request merged**
